### ISSUE TITLE
Open Grant Proposal: Freenet 2 initial prototype

### ISSUE THREAD

**Issue by sanity** [filecoin-project/devgrants#345]
# Open Grant Proposal: Freenet 2 initial prototype

**Name of Project:** Freenet 2

**Proposal Category:** `app-dev`

**Proposer:** [sanity](https://github.com/sanity)

**Technical Sponsors:** Max Inden, Dietrich Ayala

**Do you agree to open source all work you do on behalf of this RFP and dual-license under MIT and APACHE2 licenses?:** Yes

# Project Description

Freenet2 is a distributed, decentralized, key-value store in which keys are cryptographic contracts which determine what values are valid under that key. 

The store is [observable](https://en.wikipedia.org/wiki/Observer_pattern), allowing applications built on Freenet2 to listen for changes to values and be notified immediately. The cryptographic contracts are specified in web assembly.

This key-value store serves as a foundation for decentralized, scalable, and trustless alternatives to centralized services, including email, instant messaging, and social networks, many of which rely on closed proprietary protocols.

Freenet2 is implemented in Rust on top of the libp2p library and will be available across all major operating systems, Desktop and mobile.

## Value

The Internet has grown increasingly centralized over the past 25 years, such that a handful of companies now effectively control the Internet infrastructure. This happened because building decentralized applications is much more difficult than building their centralized equivalents.

Freenet2's purpose is to make it easy to build decentralized services and also to provide reference implementations of the most popular services such as email and instant messaging.

We're building Freenet2 on the Rust implementation of libp2p meaning our developers will contribute improvements to this project, including code, testing, and more general feedback. benefitting systems built on libp2p, including FileCoin. 

Our "chainless" smart contract approach will result in learnings for IPFS-native applications and FileCoin's smart contract initiative.

Lastly, our deployment of a libp2p application to the major mobile platforms will inform other IPFS on mobile efforts, particularly the low-bandwidth, low-power, and variable connectivity challenges of mobile.

### Risks

This is an ambitious project, network effects can make it very difficult to compete with centralized incumbents. However, our team has a track record of disruptive innovation and there has never been greater public concern about centralized control of the Internet. 

## Deliverables

Working software for Windows, Mac, and Linux on Desktop, and on Android and IOS - with a convenient installer. The software will support decentralized email out-of-the-box while providing a platform for future applications.

## Development Roadmap

| Milestone | Date               | Deliverables                                                                                        | Cost    |
|-----------|--------------------|-----------------------------------------------------------------------------------------------------|---------|
| 1         | Nov 30th, 2021     | Low-level networking & infrastructure, Ring topology. Key-value store. Get/put/subscribe protocols.          | $20,000 |
| 2         | January 31st, 2022 | Web Assembly Contracts, WebSocket API. Web interface. Decentralized email backend. | $30,000 |
| 3         | March 31st, 2022   | Email gateway. Email web interface.                                                                   | $15,000 |
| 4         | May 31st, 2022     | Packaging and launch for Desktop. Packing and launch for mobile                                      | $10,000 |

## Total Budget Requested

*Total Cost:* $75,000. We can accept grants in FIL.

## Maintenance and Upgrade Plans

Our plan is to continue growing our team while raising funding both from organizations and from individuals by offering
@freenet.org email accounts in return for recurring donations. If successful these will fund future development indefinitely.

# Team

## Team Members

Ian Clarke
Nacho Duart

## Team Member LinkedIn Profiles

https://www.linkedin.com/in/iancjclarke/

## Team Website

*Source code* https://github.com/freenet/locutus
*Technical design:* https://github.com/freenet/locutus/issues

### Ian Clarke

Ian designed [Freenet](https://en.wikipedia.org/wiki/Freenet), the first distributed, decentralized peer-to-peer network, first released in March 2000. Freenet pioneered many concepts, including small-world networks, cryptographic contracts, and distributed web-of-trust.

In 2003 Ian co-founded Revver, which by 2005 was the 4th most popular online video website, and the first to share
advertising revenue with video creators. As CTO, Ian recruited an engineering team of 10 developers.

Ian has also worked extensively on distributed systems including the [Swarm framework](http://swarmframework.org/), and other areas including [applying](https://vimeo.com/137999578) machine learning to online advertising.

As CTO of OneSpot, Ian recruited and managed teams of 12 developers to build a sophisticated ad targeting engine which 
outperformed Google Adwords in head-to-head comparisons.

* https://en.wikipedia.org/wiki/Ian_Clarke_(computer_scientist)
* https://github.com/sanity/

### Ignacio Duart

Ignacio Duart has been doing software engineering professionally for 7 years, lately on traditional “big data” systems, and has been a Open Source contributor for around 5 years.

In particular relevant to this project, has been doing Open Source systems programming in Rust since around 2015 (1.0 release) and has extensive familiarity with its ecosystem. Experience spans different areas: distributed systems and computing (main contributor of a distributed computing framework, Vega, based on Apache Spark), agent-based architectures and p2p network models (working on an experimental simulation framework, SimAg), work on scientific computing and ML framework implementation and bindings (eg. Tensorflow Rust bindings, implementation and ports of the GNU Scienfitic Library), familiarity with webassembly, database and compiler design amongst other things working on a query language and execution engine implemented in Rust.

* https://github.com/iduartgomez

## Team code repositories

https://github.com/freenet/

# Additional Information

**Comment by mishmosh** [comment]
Hello @sanity and thank you for this submission! We are very interested in your proposal and have a few questions and suggestions below.

1. Do you expect Freenet2 to focus on apps & use cases that need streaming/observable data stores? Or is there a world in which this might also extend to more static DB needs? Asking because there's an increasing number of application developers seeking decentralized DBs compatible with IPFS & Filecoin, and a dearth of great options for them. (Maybe @autonome @mxinden can also speak to this.) 

2. Can you expand on your plans for developer adoption of Freenet2? (This is feedback we give to 75% of applications.) A thorough [README](https://github.com/RichardLitt/standard-readme#background) with both background and usage information, excellent [docs](https://docs.web3.storage/#quickstart) with concepts, examples, and references, or video demos & tutorials would all go a long way. Once this project is ready to use, are you game for giving a talk/demo/workshop at events like hackathons, conferences, or the IPFS meetup?

3. On an administrative note, can you rework the milestones into a [table](https://www.tablesgenerator.com/markdown_tables) with roughly 3-4 milestones? Would simplify the contracting and payments process.

1 is merely a discussion question and not a blocker for evaluation, while 2 & 3 would be great to see as edits to your initial comment.

Tagging @autonome @mxinden for any additional comments or guidance, and @realChainLife who will include this in our September batch of grant evaluations.

**Comment by sanity** [comment]
@mishmosh Thank you for the quick feedback!

> 1. Do you expect Freenet2 to focus on apps & use cases that need streaming/observable data stores? Or is there a world in which this might also extend to more static DB needs? Asking because there's an increasing number of application developers seeking decentralized DBs compatible with IPFS & Filecoin, and a dearth of great options for them. (Maybe @autonome @mxinden can also speak to this.)

Freenet 2's key-value store will be able to approximate the functionality of almost any database, which will include static databases (by which I assume you mean databases that don't require real-time observability). In principle, a SQL-like layer could be created on top of the key-value store although this is not part of our MVP.

> 2. Can you expand on your plans for developer adoption of Freenet2? (This is feedback we give to 75% of applications.) A thorough README with both background and usage information, excellent docs with concepts, examples, and references, or video demos & tutorials would all go a long way. Once this project is ready to use, are you game for giving a talk/demo/workshop at events like hackathons, conferences, or the IPFS meetup?

Absolutely, the broader Freenet project has over 20 fairly active developers, among other contributors both in academia and government. We'll be taking full advantage of these relationships as the project progresses. 

Additionally, one of the many benefits of Rust is that there is a lot of developer excitement around the language which is a big help in generating interest, and we will work to build awareness and interest in Freenet 2 in popular forums like the [Rust subreddit](https://reddit.com/r/rust).

We will flesh out the repository README and related Github documentation, and also create YouTube videos for those who prefer to consume information that way. I will be more than happy to participate in events like hackathons and the IPFS meetup, as will other developers I'm sure.

I've spent most of my career building and managing engineering teams - both commercial and voluntary open source - and I intend to put these skills to full use with Freenet 2.

> On an administrative note, can you rework the milestones into a table with roughly 3-4 milestones? Would simplify the contracting and payments process.

3. No problem, I've simplified the milestones section and put them in a table, please let me know if this is satisfactory.

Please don't hesitate to let me know if you have further questions or suggestions.

**Comment by autonome** [comment]
Worked w/ Ian on this proposal and will be co-admin w/ Max - the benefits from my perspective:

* Expanding Rust libp2p community w/ this team's explicit support
* Rust libp2p on mobile is def something we're interested in, want to see perf/stability/best-practices worked out
* Very interested in chain-free computation approaches to inform future application models on IPFS

**Comment by realChainLife** [comment]
Thanks @mishmosh and @autonome 

@sanity Thanks for submitting this proposal. We would like to fund the work outlined here.

**Comment by sanity** [comment]
That's great news @realChainLife!  What are the next steps?

**Comment by jimpick** [comment]
Wow! I spent a lot of time playing around with Freenet back around 2004. The skills I learned back then taught me how to deal with the data structures used in IPFS!

**Comment by sanity** [comment]
Hi all, quick update. 

We've been making [good progress](https://github.com/freenet/locutus) over the past month, and have grown our team during this time.

We're not going to hit the estimated date for Milestone 1 of Nov 30th because we ended up tackling things in a slightly different order that made more sense once we got our teeth into things, but we expect to hit this milestone in December.

We've also been happy to see the progress on libp2p in NAT traversal over the past few weeks.

cc: @realChainLife @autonome @mxinden
